edge/pkg/devicetwin/dtmanager/twin_test.go

```diff
@@ -0,0 +1,1534 @@
+		dealType         int
+		errWant          error
+		filterReturn     orm.QuerySeter
+		insertReturnInt  int64
+		insertReturnErr  error
+		deleteReturnInt  int64
+		deleteReturnErr  error
+		updateReturnInt  int64
+		updateReturnErr  error
+		allReturnInt     int64
+		allReturnErr     error
+		queryTableReturn orm.QuerySeter
+	}{
+		{
+			name:     "TestDealDeviceTwin(): Case 1: DeviceID does not exist",
+			context:  &contextDeviceB,
+			msgTwin:  msgTwin,
+			dealType: 0,
+			errWant:  errors.New("Update rejected due to the device is not existed"),
+		},
+		{
+			name:     "TestDealDeviceTwin(): Case 2: msgTwin is nil",
+			context:  &contextDeviceB,
+			deviceID: "DeviceB",
+			dealType: 0,
+			errWant:  errors.New("Update twin error, the update request body not have key:twin"),
+		},
+		{
+			name:             "TestDealDeviceTwin(): Case 3: dealTwinResult error",
+			context:          &contextDeviceA,
+			deviceID:         "DeviceB",
+			msgTwin:          msgTwinValue,
+			dealType:         0,
+			errWant:          errors.New("The value type is not allowed"),
+			filterReturn:     mockQuerySeter,
+			allReturnInt:     int64(1),
+			allReturnErr:     nil,
+			queryTableReturn: mockQuerySeter,
+		},
+		{
+			name:             "TestDealDeviceTwin(): Case 4: DeviceTwinTrans error",
+			context:          &contextDeviceB,
+			deviceID:         "DeviceB",
+			msgTwin:          msgTwin,
+			dealType:         0,
+			errWant:          errors.New("Failed DB Operation"),
+			filterReturn:     mockQuerySeter,
+			insertReturnInt:  int64(1),
+			insertReturnErr:  errors.New("Failed DB Operation"),
+			deleteReturnInt:  int64(1),
+			deleteReturnErr:  nil,
+			updateReturnInt:  int64(1),
+			updateReturnErr:  nil,
+			allReturnInt:     int64(1),
+			allReturnErr:     nil,
+			queryTableReturn: mockQuerySeter,
+		},
+		{
+			name:             "TestDealDeviceTwin(): Case 5: Success Case",
+			context:          contextDeviceC,
+			deviceID:         "DeviceC",
+			msgTwin:          msgTwin,
+			dealType:         0,
+			errWant:          nil,
+			filterReturn:     mockQuerySeter,
+			allReturnInt:     int64(1),
+			allReturnErr:     nil,
+			queryTableReturn: mockQuerySeter,
+		},
+	}
+
+	fakeDevice := new([]dtclient.Device)
+	fakeDeviceArray := make([]dtclient.Device, 1)
+	fakeDeviceArray[0] = dtclient.Device{ID: "DeviceB"}
+	fakeDevice = &fakeDeviceArray
+
+	fakeDeviceAttr := new([]dtclient.DeviceAttr)
+	fakeDeviceAttrArray := make([]dtclient.DeviceAttr, 1)
+	fakeDeviceAttrArray[0] = dtclient.DeviceAttr{DeviceID: "DeviceB"}
+	fakeDeviceAttr = &fakeDeviceAttrArray
+
+	fakeDeviceTwin := new([]dtclient.DeviceTwin)
+	fakeDeviceTwinArray := make([]dtclient.DeviceTwin, 1)
+	fakeDeviceTwinArray[0] = dtclient.DeviceTwin{DeviceID: "DeviceB"}
+	fakeDeviceTwin = &fakeDeviceTwinArray
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if reflect.DeepEqual(test.name, "TestDealDeviceTwin(): Case 3: dealTwinResult error") {
```

Comment:
this is looking a little ugly . can we do better ? maybe we can breakup the tests and avoid this if-else conditions check on the test case name.